### ISSUE TITLE
pyplot import

### ISSUE THREAD

**Issue by RMeli** [swcarpentry/python-novice-inflammation#830]
In the episode [Visualizing Tabular Data](https://swcarpentry.github.io/python-novice-inflammation/03-matplotlib/index.html), `pyplot` is imported as follows:

```
import matplotlib.pyplot
```

This import requires to use `matplotlib.pyplot` everywhere in the code, which is much longer than the usual convention `plt`. I suggest to change the import to

```
import matplotlib.pyplot as plt
```
or (see Issue #665)
```
from matplotlib import pyplot as plt
```

and to update the code accordingly. For example the first cell would become
```
from matplotlib import pyplot as plt
image = plt.imshow(data)
plt.show()
```

which is much easier to read (the emphasis is on the function being called and not `matplotlib.pyplot`) and introduces the `plt` name which is pretty standard.

**Comment by ldko** [comment]
Hi @RMeli , thank you for taking time to contribute this suggestion. You are right that using `as plt` would be less to type and follows a common practice.

Using `as` to supply a name for an imported module is something that has come up in discussion before, enough that it has found a place on our short list of ["Frequently Argued Issues"](https://swcarpentry.github.io/python-novice-inflammation/guide/). This links to a discussion that happened on this topic in pull request #61. The decision was to keep the import as `import matplotlib.pyplot` to keep from overloading learners new to programming with an extra concept. It is left up to an instructor teaching the lesson to decide if they feel teaching the `as` syntax would be right for their audience. There is a callout, entitled "Importing libraries with shortcuts",  in the [Analyzing Patient Data episode](https://swcarpentry.github.io/python-novice-inflammation/02-numpy/index.html) that mentions this style of import with `import numpy as np` for those who want to apply it.

**Comment by RMeli** [comment]
@ldko Thanks for pointing out #61 and the "Frequently Argued Issues" , which I obviously missed. 

I think the cognitive load of "you can use `as` to rename your import, which is useful for shortcuts" is very low compared to the benefits; my vote would have contribute to a 50/50 split on the matter! ;)

I'm happy to close the issue following #61.

**Comment by ldko** [comment]
> I think the cognitive load of "you can use as to rename your import, which is useful for shortcuts" is very low compared to the benefits; my vote would have contribute to a 50/50 split on the matter! ;)

Your point is certainly valid. People feel pretty strongly one way or another which is why I believe it made the Frequently Argued Issues, and we leave it to the instructors to decide. There are many nuances and tips to teach to people learning Python; we find ourselves having to draw lines in the lesson text on what to include and leave out in order to keep the amount of content manageable for a workshop time frame and for those brand new to programming.

Please feel free to close the issue. Thank you again, @RMeli !

**Comment by maxim-belkin** [comment]
Hi, Rocco! 👋 
Cognitive load is the amount of information learners have to store in their short-term memory to understand the topic. That short-term memory is limited. The benefit you're talking about are related to the amount of typing one has to do to complete a task. If we fail to make learners understand the topic, they won't be able to use the benefits you mention.

**Comment by RMeli** [comment]
I completely understand. Thank you for your comments @ldko !

**Comment by RMeli** [comment]
@maxim-belkin thank you for your comments. I completely understand, but I don't think the benefits are limited to the amount of typing. Using a shortcut for long imports puts the emphasis on the function being called. I think for a student
```
matplotlib.pyplot.show()
```
is much more difficult to digest than
```
plt.show()
```
In the second example the function call is clear, simple, and easy to read. In the first example, the student has to read `matplotlib.pyplot` before reading the actual function call. This makes the call less clear and could spark some questions (for example, why there is a `.` between `matplotlib` and `pyplot` as well as between `pyploy` and `show`). The answer tu such questions would probably increase the cognitive load even further.

Another benefit is that `as` is very common and widely used in any online tutorial or actual script; if students look up for something on the internet, they will likely encounter `as` immediately.

However, this is my opinion and from #61 it seems pretty divisive. Since discussing this all over again would not serve any purpose (I think we would be getting another 50/50 split) I closed the issue.

**Comment by RMeli** [comment]
I'm re-opening this issue because I realized that import shortcuts are introduced in the previous lesson: [Analyzing Patient Data](https://swcarpentry.github.io/python-novice-inflammation/02-numpy/index.html).

If the concept is introduced in a lesson, what are the arguments against using it in subsequent lessons?

**Comment by ldko** [comment]
Hi @RMeli , You are correct, we do bring up shortcut style imports in the Analyzing Patient Data episode (mentioned [earlier in this thread](https://github.com/swcarpentry/python-novice-inflammation/issues/830#issuecomment-640671464)) in the "Importing libraries with shortcuts" callout box. There it is presented as an optional way of doing imports, and it suggests that using shortcuts or not is for the individual to decide on, but that they should be consistent. We have decided not to use the `as` syntax in the lesson material beyond presenting it as an option. We do not mandate that instructors in a workshop or people using the lesson materials on their own avoid using the shortcut :). Regarding:

```
matplotlib.pyplot.show()
```

vs.

```
plt.show()
```
and your comment:

> In the second example the function call is clear, simple, and easy to read. In the first example, the student has to read matplotlib.pyplot before reading the actual function call. This makes the call less clear and could spark some questions (for example, why there is a . between matplotlib and pyplot as well as between pyploy and show?). The answer to such questions would probably increase the cognitive load even further.

I see that `plt.show()` could seem more simple to some since there is only one `.` as you point out. I think it is debatable as to whether or not it is more clear than `matplotlib.pyplot.show()`. A learner will still see the `matplotlib.pyplot` in `import matplotlib.pyplot as plt`and may similarly wonder about the `.` between `matplotlib` and `pyplot` there as well. While importing `numpy as np` and `matplotlib.pyplot as plt` may be common for those particular modules, `as` imports are not as commonly used for many other modules, so seeing the imports without `as` can be helpful in the lesson so as not to imply that one always needs an `as` in importing.

More to my point, `plt`, while a common convention, is an arbitrary name being given to `matplotlib.pyplot`, so it could look kind of magic as to why `plt.show()` works to a beginner. Using the full name reference has meaning and shows that there is a function `show` in a module called `matplotlib.pyplot`. The dotted notation used here is taught just before the "Importing libraries with shortcuts" callout box. Explaining `plt.show()` for me would mean saying `plt` is an arbitrary name we gave a module called `matplotlib.pyplot` from which we are using the module `show()`. I perceive this as an extra layer to be taught and understood.

**Comment by jack89roberts** [comment]
I understand this has previously been argued and decided so sorry for adding more, but perhaps adding a short note that `import matplotlib.pyplot as plt` is a common convention would be worthwhile (similar to the "Importing libraries with shortcuts" box in the previous episode that @RMeli mentioned, though here it could be a one-liner as it's already been introduced previously)?

I'd also slightly counter that it's only a common convention - it's _the_ convention for numpy and matplotlib. To the extent that if you look at the documentation and examples returned by `help(numpy.mean)` ([or on the web-site](https://numpy.org/doc/stable/reference/generated/numpy.mean.html)) they show using `np.mean` without even including the `import numpy as np` before - they assume you've already done that.

**Comment by ldko** [comment]
I think a short note about the `plt` style import would be alright to add if you want to open a pull request. Thanks @jack89roberts

**Comment by jack89roberts** [comment]
Great, thanks @ldko - I've had a first go at it in #910

**Comment by RMeli** [comment]
Addressed by #910
